x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -344,6 +351,11 @@ public static IndexResolution mergedMappings(DataTypeRegistry typeRegistry, Stri
                 }
             }
 
+            // if there are both a keyword and a constant_keyword type for this field, only keep the keyword as a common compatible type
+            if (hasCompatibleKeywords) {
+                types.remove(CONSTANT_KEYWORD.esType());
```

Comment:
Removing the type all together might remove useful information.
For example if a field a is mapped as keyword in index X and constant_keyword in index Y, the piece above will find the collision and remove the field from Y resulting in a mapping only in X.
What happens is there's no removal?

Reply:
This scenario is the one where two indices have the **same named field** with both `keyword` and `constant_keyword` types and their mappings need to be merged. For other scenarios (where the same name is used for two fields with different types) we tell the user that the field cannot be shown (because of the different mapping types), but in this very specific case and since these two field types are so similar I chose to display the field with the "common" data type as `keyword`.

The controversy for what to do in this specific case came from the scenario described [here](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/faster-filtering-with-constant-keyword.html) and is specifically tested [here](https://github.com/elastic/elasticsearch/pull/53241/files#diff-130379dcaa8e58529d8c41a04cf85c10R274).